# ==============================================================================
# FIXING COMMIT DIFF
# ==============================================================================
# File: dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp
# Commit: f3d36f83998f
# Full Hash: f3d36f83998f4e563c29e0ac545bec4a882bfaec
# Author: Kagami Sascha Rosylight <saschanaz@outlook.com>
# Date: 2025-03-06 21:30:30
# Description:
#   Bug 1951956 - Clean up MozPromiseHolder when shutting down r=asuth
#   
#   The callback may never be called if the parent process shuts down before getting the async result.
#   
#   Another reason why bug 1937194 is wanted.
# ==============================================================================

diff -r 6efa50f6acf0 -r f3d36f83998f dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp
--- a/dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp	Thu Mar 06 08:44:13 2025 +0000
+++ b/dom/serviceworkers/ServiceWorkerRegistrationProxy.cpp	Thu Mar 06 09:00:24 2025 +0000
@@ -515,7 +515,11 @@
   }
 
  protected:
-  virtual ~NotificationsCallback() = default;
+  virtual ~NotificationsCallback() {
+    // We may be shutting down prematurely without getting the result, so make
+    // sure to settle the promise.
+    mPromiseHolder.RejectIfExists(NS_ERROR_DOM_INVALID_STATE_ERR, __func__);
+  };
 
   MozPromiseHolder<NotificationsPromise> mPromiseHolder;
 };
